### ISSUE TITLE
Using postfix operators for chained function calls doesn't work?

### ISSUE THREAD

**Issue by myrklewk** [benjamin-hodgson/Pidgin#11]
I've been trying to implement chained function calls, i.e. foo(bar)(baz)
It seemed like the obvious way to do this is to use a postfix operator but I haven't managed to get it to work. The first call will parse (foo(bar)), but the operator won't chain. I looked at the source for the expression builder, and it looks like, unlike with binary operators, unary operators are not recursively applied and aggregated.

Should the unary operators be recursively applied? Can you suggest how I should get this to work?

In your conversation with kswaroop1 re: his "expression example" pull request you suggest doing exactly this for function call operators (including the foo(bar)(baz) example), but there isn't enough of an example for me to understand how to make this work.

**Comment by benjamin-hodgson** [comment]
Thanks for your question. Please could you post an [MCVE](https://stackoverflow.com/help/mcve)?

**Comment by myrklewk** [comment]
The example below prints:

[call foo]
[add [add aa to bb] to cc]

I would like the first line to be instead: [call [call foo]]

program.cs
```
using System;
using Pidgin;

namespace pidgin_test
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine(ParserBuilder.Expression.Parse("foo()()").Value);
            Console.WriteLine(ParserBuilder.Expression.Parse("aa+bb+cc").Value);
        }
    }
}
```
ParserBuilder.cs
```
using System;
using Pidgin;
using static Pidgin.Parser;
using static Pidgin.Parser<char>;
using Pidgin.Expression;

namespace pidgin_test
{
    public class Node {}

    public class IdentifierNode : Node
    {
        private readonly string name_;

        public IdentifierNode(string name)
        {
            name_ = name;
        }

        public override string ToString() => name_;
    }

    public class CallNode : Node
    {
        private readonly Node iden_;

        public CallNode(Node iden)
        {
            iden_ = iden;
        }

        public override string ToString() => $"[call {iden_.ToString()}]";
    }

    public class AddNode : Node
    {
        private readonly Node l_, r_;

        public AddNode(Node l, Node r)
        {
            l_ = l;
            r_ = r;
        }

        public override string ToString() =>$"[add {l_.ToString()} to {r_.ToString()}]";
    }

    public static class ParserBuilder
    {
        public static readonly Parser<char, string> Identifier =
            from first in Token(char.IsLetter)
            from tail in Token((c) => char.IsLetterOrDigit(c) || '_' == c).ManyString()
            select first + tail;

        public static readonly Parser<char, Node> IdenNode =
            Identifier.Select(n => (Node)new IdentifierNode(n));

        public static readonly Parser<char, Node> Expression =
            ExpressionParser.Build(IdenNode,
                new[]
                {
                    Operator.Postfix(Char('(').Then(Char(')').Select<Func<Node, Node>>(_ => n => new CallNode(n)))),
                    Operator.InfixL(Char('+').ThenReturn<Func<Node, Node, Node>>((l, r) => new AddNode(l, r)))
                });
    }
}

```

**Comment by benjamin-hodgson** [comment]
Ah, I understand your question. The expression parser doesn't chain prefix/postfix operators by default, because that's not always what you want (`- - 1`, "negative negative 1", is not valid syntax in many languages).

The simplest fix in your case is to treat a _chain_ of function invocations as a single postfix operator, so...

```csharp
Operator.Postfix(
    String("()")
        .AtLeastOnce()
        .Select<Func<Node, Node>>(xs => x => xs.Aggregate(x, (acc, _) => new CallNode(acc)))
);
```

Let me know whether that works for you.

**Comment by myrklewk** [comment]
Ahhh, yes. That looks to be exactly what I want. I will give it a try (it would have taken me forever to realize I could do the aggregation in the postfix operator!).

**Comment by benjamin-hodgson** [comment]
The prefix equivalent would need to use a right associative fold.

**Comment by myrklewk** [comment]
Makes sense now that I see it :-)
Worked perfectly, though the helper would make it much cleaner. Thanks!

**Comment by benjamin-hodgson** [comment]
You can define your own helper in the mean time 😉

**Comment by benjamin-hodgson** [comment]
Released in v1.5.0 https://www.nuget.org/packages/Pidgin
